config_builder.go

```diff
@@ -1072,65 +1083,69 @@ func (d *DefaultDatabaseBuilder) BuildDatabase(
 	case err != nil:
 		cleanUp()
 
-		err := fmt.Errorf("unable to open graph DB: %w", err)
+		err = fmt.Errorf("unable to open graph DB: %w", err)
 		d.logger.Error(err)
 		return nil, nil, err
 	}
 
-	// Instantiate a native SQL invoice store if the flag is set.
+	// Instantiate a native SQL store if the flag is set.
 	if d.cfg.DB.UseNativeSQL {
-		// We need to apply all migrations to the native SQL store
-		// before we can use it.
-		err := dbs.NativeSQLStore.ApplyAllMigrations(
-			sqldb.GetMigrations(),
-		)
-		if err != nil {
-			cleanUp()
-			err := fmt.Errorf("unable to apply migrations: %w", err)
-			d.logger.Error(err)
-
-			return nil, nil, err
-		}
+		migrations := sqldb.GetMigrations()
+
+		// If the user has not explicitly disabled the SQL invoice
+		// migration, attach the custom migration function to invoice
+		// migration (version 6). Even if this custom migration is
+		// disabled, the regular native SQL store migrations will still
+		// run. If the database version is already above this custom
+		// migration's version (6), it will be skipped permanently,
+		// regardless of the flag.
+		if !d.cfg.DB.SkipSQLInvoiceMigration {
+			migrationFn := func(tx *sqlc.Queries) error {
+				return invoices.MigrateInvoicesToSQL(
+					ctx, dbs.ChanStateDB.Backend,
+					dbs.ChanStateDB, tx,
+					invoiceMigrationBatchSize,
+				)
+			}
 
-		// KV invoice db resides in the same database as the channel
-		// state DB. Let's query the database to see if we have any
-		// invoices there. If we do, we won't allow the user to start
-		// lnd with native SQL enabled, as we don't currently migrate
-		// the invoices to the new database schema.
```

Comment:
dont we still want this check for anyone who _has_ set `d.cfg.DB.SkipSQLInvoiceMigration=true`?

Reply:
and to protect against users who had a bbolt invoice store before?

Reply:
ah yes i see: if the user was pointing to a bbolt store and then changes config to sql store, there really isnt any way for us to know this so it is essentially a new node